Sprint 16

bakgeldia opened the pull request:
1. Экран со списком категорий переписан на архитектуру MVVM  с реализацией байндингов через замыкания. 
2. Сделаны экраны онбординга с использованием UIPageViewController.

bakgeldia committed:
Добавлен экран онбординга с использованием UIPageViewController и UIPageControl.

bakgeldia committed:
Экран со списком категорий переписан на архитектуру MVVM с реализацией байндингов через замыкания

bakgeldia committed:
Реализована функция при нажатии на кнопку на экране онбординга: в следующей открытии приложения он не будет показываться

bakgeldia committed:
Удален лишний комментарий

romanenko-ivan commented on Tracker/Trackers/TrackerCreation/HabitCreation/Сategory/CategoryViewController.swift:
```diff
@@ -15,16 +15,37 @@ final class CategoryViewController: UIViewController {
     
     weak var delegate: CategoryViewControllerDelegate?
     
+    private let trackerCategoryStore = TrackerCategoryStore()
```
Хоть свойство и не используется, но для соблюдения архитектуры давай его удалим

romanenko-ivan commented on Tracker/Trackers/TrackerCreation/HabitCreation/Сategory/CategoryTableViewCell.swift:
```diff
@@ -0,0 +1,46 @@
+    func configure(with title: String, isSelected: Bool) {
+        titleLabel.text = title
+        accessoryType = isSelected ? .checkmark : .none
+        backgroundColor = UIColor(red: 230.0/255.0, green: 232.0/255.0, blue: 235.0/255.0, alpha: 0.3)
```
Такой код трудно переиспользовать, давай  расширим `Color` вынесем этот код в отдельное свойство, так будет легче переиспользовать и понять по названию о каком цвете идет речь

romanenko-ivan commented on Tracker/ViewController.swift:
```diff
@@ -8,13 +8,27 @@
 import UIKit
 
 class ViewController: UITabBarController {
```
Давай этому классу дадим более предметное название, например `TabBarController`

romanenko-ivan commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -60,9 +60,8 @@ final class OnboardingViewController: UIPageViewController, UIPageViewController
-        let viewController = ViewController()
-        viewController.modalPresentationStyle = .fullScreen
-        present(viewController, animated: false, completion: nil)
+        UserDefaults.standard.set(true, forKey: "hasSeenOnboarding")
```
Давай вынесем ключ в константы

bakgeldia committed:
Исправление после код-ревью №1:
- Удален неиспользуемый экземпляр класса TrackerCategoryStore из CategoryViewController
- На экранах создания трекера добавлена возможность скрыть клавиатуру при тапе на любом месте экрана

bakgeldia committed:
Изменения после код-ревью №2:
- ViewController переименен на TabBarController
- Ключ для проверки просмотра пользователем онбординга вынесен в константу onboardKey
- Цвета вынесены в отдельную структуру Colors

bakgeldia committed:
Код переорганизован по структуре:
Публичные переменные
IBOutlet
Приватные переменные
Публичные методы(сначала initы, overrideы, и потом остальные)
Приватные методы
IBAction

bakgeldia committed:
Повторяющиеся коды OnboardingFirstPageViewController и OnboardingSecondPageViewController объединены в OnboardingPageViewController. Создан enum OnboardingPageType, который используется при конфигурировании названия картинки и текста

romanenko-ivan commented:
Отличная работа!

🍪 Для размышлений 

Куда дальше: MVVM - архитектура, которая требует использования байндинговых взаимосвязей и наилучшим образом работает с реактивными фреймворками.
Чаще всего в таких случаях используются RxSwift и иногда ReactiveSwift. 
Эти два фреймворка часто встречаются в проектах, которые еще не перешли на Combine - нативный фреймворк от Apple.
Я хочу отметить, что при выборе этой архитектуры для тестовых заданий при трудоустройстве использование одного из этих фреймворков будет очень сильным показателем.

Ссылки на соответствующие репозитории и статьи:

- [RxSwift](https://github.com/ReactiveX/RxSwift)
- [Пример использования MVVM с бесконечной прокруткой изображений на Swift](https://betterprogramming.pub/mvvm-in-swift-infinite-scrolling-and-image-loading-d47780b06e23)
- [ReactiveSwift](https://github.com/ReactiveCocoa/ReactiveSwift/)
- [Пример реализации MVVM с использованием ReactiveCocoa](https://blog.joanzapata.com/mvvm-reactivecocoa-5/)
- [Документация Apple по фреймворку Combine](https://developer.apple.com/documentation/combine)
